### ISSUE TITLE
Allocate FUSE call reserves 18 bytes too much disk space

### ISSUE THREAD

**Issue by slackner** [rfjakob/gocryptfs#311]
Steps to reproduce:
```
fallocate -l 8192 testfile
```

The debug output shows:
```
Allocate off=0 sz=8192 mode=0 cipherOff=18 cipherSz=8274
```

A reservation of 8274 byte is expected (i.e., two blocks with 4092+32 byte, and a 18 byte header), however, since `cipherOff != 0`, this effectively reserves 18 bytes too much.

**Comment by slackner** [comment]
I would also argue that allocating a region in the middle of the file should always reserve full blocks since partial blocks are not allowed there.

**Comment by rfjakob** [comment]
You mean a reservation of 8266 bytes is expected?

I checked with strace, we reserve 8274 bytes, as you said:

> [pid  6423] ftruncate(14, 8274)         = 0

**Comment by slackner** [comment]
The problem seems to be here:

```
	cipherSz := lastBlock.BlockCipherOff() - cipherOff +
		f.contentEnc.PlainSizeToCipherSize(lastBlock.Skip+lastBlock.Length)
```

The function `f.contentEnc.PlainSizeToCipherSize` also adds the 18 bytes header, but this would not be required here. I hope it is now more clear what I mean.

Nevertheless, even with that fixed, the function should behave differently if blocks are somewhere in the middle of the file. Reserving 1 byte of an inner block (of size 4096) should reserve the full block (since it will be required anyway as soon as we write to the file). Otherwise there is no guarantee that the write will succeed.

**Comment by rfjakob** [comment]
Ah, here it is, missed it in the strace output:
> [pid  6423] fallocate(14, FALLOC_FL_KEEP_SIZE, 18, 8274) = 0

**Comment by slackner** [comment]
Another corner case that is not handled correctly yet: If the previous file size is zero, we should also explicitly fallocate the first 18 bytes for the header.

**Comment by rfjakob** [comment]
Merged as 40d2427fd7f788262dc3bf220f789b7bd5fe781b , thanks
